samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs

```diff
@@ -0,0 +1,384 @@
+        for (var i = 0; i < nodes.Count; ++i)
+        {
+            var node = nodes[i];
+            var functionName = node!.LocalName;
+            var skillName = string.Empty;
+
+            if (s_skillMapping.TryGetValue(node!.LocalName, out KeyValuePair<string, string> value))
+            {
+                functionName = value.Value;
+                skillName = value.Key;
+            }
+
+            var hasChildElements = node.HasChildElements();
+
+            if (hasChildElements)
+            {
+                plan.AddSteps(NodeListToPlan(node.ChildNodes, context, functionName));
+            }
+            else
+            {
+                if (
+                    string.IsNullOrEmpty(skillName) ? !context.Skills!.HasFunction(functionName) : !context.Skills!.HasFunction(skillName, functionName))
+                {
+                    var planStep = new Plan(node.InnerText);
+                    planStep.NamedParameters.Update(node.InnerText);
+                    planStep.NamedOutputs.Set($"markup.{functionName}.result", string.Empty);
+                    plan.NamedOutputs.Set($"markup.{functionName}.result", string.Empty);
+                    plan.AddSteps(planStep);
+                }
+                else
+                {
+                    var command = string.IsNullOrEmpty(skillName)
+                        ? context.Skills.GetFunction(functionName)
+                        : context.Skills.GetFunction(skillName, functionName);
+                    var planStep = new Plan(command);
+                    planStep.NamedParameters.Update(node.InnerText);
+                    planStep.NamedOutputs.Set($"markup.{functionName}.result", string.Empty);
+                    plan.NamedOutputs.Set($"markup.{functionName}.result", string.Empty);
+                    plan.AddSteps(planStep);
+                }
+            }
+        }
+
+        return plan;
+    }
+}
+
+#region Utility Classes
+
+public class XmlMarkup
+{
+    public XmlMarkup(string response, string? wrapperTag = null)
+    {
+        if (!string.IsNullOrEmpty(wrapperTag))
+        {
+            response = $"<{wrapperTag}>{response}</{wrapperTag}>";
+        }
+
+        this.Document = new XmlDocument();
+        this.Document.LoadXml(response);
+    }
+
+    public XmlDocument Document { get; }
+
+    public XmlNodeList SelectAllElements()
+    {
+        return this.Document.SelectNodes("//*")!;
+    }
+
+    public XmlNodeList SelectElements()
+    {
+        return this.Document.SelectNodes("/*")!;
+    }
+}
+
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+public struct XmlNodeInfo
+{
+    public int StackDepth { get; set; }
+    public XmlNode Parent { get; set; }
+    public XmlNode Node { get; set; }
+
+    public static implicit operator XmlNode(XmlNodeInfo info)
+    {
+        return info.Node;
+    }
+}
+#pragma warning restore CA1815
+
+#pragma warning disable CA1711
+public static class XmlEx
```

Comment:
why are we still using XML and not leveraging the plan data structure?

Reply:
This literally is a "XmlMarkup" Plan example, showing how you might create a Plan (data structure) out of your own XML. I would think XML parsing would be a required component of such an example.